### PR TITLE
[Merged by Bors] - Optimise slasher DB layout and switch to MDBX

### LINKED ISSUE #2286: Optimise slasher storage of attester data
## Description

The slasher database is still growing, but ideally we would like it to settle down to an approximately constant size. I initially thought this was a problem with pruning (#1890), but it seems that isn't the case.

Analysing a slasher DB from mainnet which has been in use since genesis 4 months ago, we see a large number of pages in use for the `attesters` sub-database. In a DB of ~95 GiB the breakdown is roughly:

- 55.4 GiB for the `attesters` (14527008 leaf pages x 4 KiB), 448M entries :scream:
- 37.8 GiB for the `indexed_attestations` (9898253 leaf pages) 29M entries
- 2 GiB for _everything else_

Shrinking the `indexed_attestations` data is the subject of https://github.com/sigp/lighthouse/issues/2112.

To shrink the `attesters` data will require a database schema change. Currently each entry is at least 80 bytes:

- 8 bytes: target epoch (key)
- 8 bytes: validator index (key)
- 32 bytes: the hash of the attestation data (value)
- 32 bytes: the hash of the indexed attestation (value)

Given this, we might naïvely expect the attesters DB to be 33 GiB (`448M x 80 bytes`) . What's the extra 22 GiB?? My current theory is that it's the previous versions of some items, which is how LMDB implements multi-version concurrency control (MVCC). There's some more info in this presentation from the LMDB creator:

http://schd.ws/hosted_files/buildstuff14/96/20141120-BuildStuff-Lightning.pdf

I quite like LMDB, and I'm willing to cop the ~2x overhead to retain its speediness. Hence I think our best option is to make that sub-database intrinsically smaller. Some ideas:

- The attestation data hash is present _only_ for speed, it could be looked up from the indexed attestation. Although this would require us to do hashing, which would be too slow for _every_ incoming attestation... so a bit of a non-starter.
- We could normalise the data so that there's a layer of indirection between the attesters DB and the 32 byte hashes. Add two new tables: `indexed_attestation_hashes` and `attestation_data_hashes` which map `u64 => Hash256`. This would shrink the two 32 byte hashes in the `attesters` DB to 16 bytes total, plus the total size of the `_hashes` tables, which should be small due to the number of attesters per attestation.
- We could use 5 bytes to encode the validator index (max validator registry size is `2**40`)
- We could use 2 bytes to encode the target epoch (store it modulo `history_length`/4096)

The last 3 of these optimisations would bring attester entries to 16 + 7  = 23 bytes, a great reduction from 80! This might yield an attesters table of max size 19.19 GiB (`448M * 23 bytes * 2 versions`). Which gives us some headroom as the validator set grows!

## Version

Lighthouse v1.2.1

## Extra Info

Complete database stats for the DB described above:

```
$ mdb_stat -a -e slasher_db
Environment Info
  Map address: (nil)
  Map size: 274877906944
  Page size: 4096
  Max pages: 67108864
  Number of pages used: 24775669
  Last transaction ID: 868787
  Max readers: 126
  Number of readers used: 0
Status of Main DB
  Tree depth: 1
  Branch pages: 0
  Leaf pages: 1
  Overflow pages: 0
  Entries: 7
Status of attesters
  Tree depth: 5
  Branch pages: 135183
  Leaf pages: 14527008
  Overflow pages: 0
  Entries: 448178532
Status of current_epochs
  Tree depth: 3
  Branch pages: 5
  Leaf pages: 727
  Overflow pages: 0
  Entries: 112640
Status of indexed_attestations
  Tree depth: 5
  Branch pages: 179246
  Leaf pages: 9898253
  Overflow pages: 0
  Entries: 28548391
Status of max_targets
  Tree depth: 3
  Branch pages: 20
  Leaf pages: 2525
  Overflow pages: 0
  Entries: 110722
Status of metadata
  Tree depth: 1
  Branch pages: 0
  Leaf pages: 1
  Overflow pages: 0
  Entries: 2
Status of min_targets
  Tree depth: 3
  Branch pages: 110
  Leaf pages: 14526
  Overflow pages: 0
  Entries: 112640
Status of proposers
  Tree depth: 3
  Branch pages: 55
  Leaf pages: 8109
  Overflow pages: 0
  Entries: 129717
```

### LINKED ISSUE #2342: LMDB slasher pre-allocates max DB size on Windows
## Description

All current releases of LMDB (up to 0.9.29) require the full size of the database to be allocated on Windows. This means that the slasher will create a 256GB file immediately upon starting!

## Details

It turns out the LMDB codebase has two main branches:

- [`mdb.RE/0.9`](https://github.com/LMDB/lmdb/tree/mdb.RE/0.9): stable code deemed worthy of release as part of the v0.9.x series
- [`mdb.master`](https://github.com/LMDB/lmdb/tree/mdb.master): experimental/under-development code

The bug we are interested in patching is [ITS#8324](https://bugs.openldap.org/show_bug.cgi?id=8324), which causes LMDB to pre-allocate the entirety of the mapped file on Windows. The bugfix is unfortunately only available on `mdb.master` at the moment, so no stable release of LMDB exists for us to target. I confirmed that using `mdb.master` resolves the issue, and the tests pass. For reference, this is the patch for `Cargo.toml`:

```toml
[patch]
[patch.crates-io]
lmdb-sys = { git = "https://github.com/michaelsproul/lmdb-rs", branch = "windows-fix-track-master" }
```

We can't use this solution in prod because we don't know what gnarly bugs may be lurking in the weeds of LMDB's dev branch. Instead we must migrate to a suitable LMDB release once one becomes available.

The release we want should include `NtCreateSection`: https://github.com/LMDB/lmdb/blob/4b6154340c27d03592b8824646a3bc4eb7ab61f5/libraries/liblmdb/mdb.c#L4488

### LINKED ISSUE #2538: Slasher Database Errors (MapFull)
## Description

Beacon Node throwing errors about slasher service and pruning of slasher database

## Version

Docker image `v1.5.0` (I also saw this in `v1.4.0`)

## Present Behaviour

I get a steady stream of errors related to the slasher service:
```
Aug 24 20:48:11.729 ERRO Error during slasher database pruning   error: DatabaseError(MapFull), epoch: 59979, service: slasher
Aug 24 20:48:25.437 ERRO Error processing array update           error: DatabaseError(MapFull), service: slasher
Aug 24 20:48:25.441 ERRO Error during scheduled slasher processing, error: DatabaseError(MapFull), epoch: 59979, service: slasher
...
Aug 24 20:48:50.141 ERRO Error during slasher database pruning   error: DatabaseError(MapFull), epoch: 59979, service: slasher
Aug 24 20:49:04.834 ERRO Error processing array update           error: DatabaseError(MapFull), service: slasher
Aug 24 20:49:04.837 ERRO Error during scheduled slasher processing, error: DatabaseError(MapFull), epoch: 59979, service: slasher
```

## Expected Behaviour

No errors?

## Steps to resolve

I have two servers, both running via docker, both with the same behavior...

Here is my sanitized `docker-compose.yaml` (`<redacted>`)
```
version: "3.4"

services:
  lh-beacon:
    container_name: lh-beacon
    image: "sigp/lighthouse:v1.5.0"
    volumes:
      - "/data/volumes/lh-beacon:/root/.lighthouse"
    ports:
      - "9000:9000" # Peering (TCP)
      - "9000:9000/udp" # Peering (UDP)
    restart: unless-stopped
    command:
      - lighthouse
      - beacon_node
      # FLAGS
      - --disable-upnp
      - --eth1
      - --http
      - --metrics
      - --slasher
      - --subscribe-all-subnets
      - --import-all-attestations
      - --validator-monitor-auto
      # OPTIONS
      - --monitoring-endpoint=<redacted>
      - --eth1-endpoints=<redacted>
      - --http-address=0.0.0.0
      - --http-allow-origin=*
      - --metrics-address=0.0.0.0
      - --metrics-allow-origin=*
      - --network=mainnet
      - --port=9000
      - --target-peers=50
      - --slasher-broadcast
      - --libp2p-addresses=<redacted>
      - --trusted-peers=<redacted>
      - --validator-monitor-pubkeys=<redacted>
    logging:
      driver: "local"
```

This is a really powerful server, plenty of RAM, plenty of disk space. I assume this has something to do with trying to run a slasher subscribed to all subnets or something? I would think this should be a possible setup?

OS is `Ubuntu 20.04.2 LTS (GNU/Linux 5.4.0-80-generic x86_64)`

```bash
# docker version
Client: Docker Engine - Community
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:54:27 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:52:33 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b63
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Closes #2286
Closes #2538
Closes #2342

## Proposed Changes

Part II of major slasher optimisations after #2767

These changes will be backwards-incompatible due to the move to MDBX (and the schema change) :scream: 

* [x] Shrink attester keys from 16 bytes to 7 bytes.
* [x] Shrink attester records from 64 bytes to 6 bytes.
* [x] Separate `DiskConfig` from regular `Config`.
* [x] Add configuration for the LRU cache size.
* [x] Add a "migration" that deletes any legacy LMDB database.

### CODE REVIEW

**Comment by michaelsproul** [comment]
I've just pushed a commit that shrinks the attester records even further to a mere 6 bytes! This means that our attesters table is now 13 bytes per entry rather than 80! :tada:

There seems to be a slight hit to batch processing time, as we now have to load indexed attestations to check for double votes in the case where we process two distinct indexed attestations for the same validator. This shouldn't occur too often anymore because of de-duplication, and I've added the `SLASHER_NUM_INDEXED_ATTESTATION_LOADS` metric to track it.

I'm going to observe the performance and storage growth over a longer period before marking this ready for review

**Comment by michaelsproul** [comment]
Threw an LRU cache in there to reduce the number of unnecessary reads. It hasn't impacted total running time as much as I thought, but it's still looking pretty low, and the slasher seems to settle into its happy path sooner after start up.

File: slasher/Cargo.toml [hunk 1]
```diff
@@ -13,8 +13,9 @@ flate2 = { version = "1.0.14", features = ["zlib"], default-features = false }
 lazy_static = "1.4.0"
 lighthouse_metrics = { path = "../common/lighthouse_metrics" }
 filesystem = { path = "../common/filesystem" }
-lmdb = "0.8"
-lmdb-sys = "0.8"
+# mdbx = { git = "https://github.com/vorot93/mdbx-rs", rev = "1150cafb38d6136c863c4618a4eff6a13723537d" }
+mdbx = { git = "https://github.com/michaelsproul/mdbx-rs", branch = "windows" }
```
Comment (michaelsproul) [inline slasher/Cargo.toml]: Will switch this over to upstream once https://github.com/vorot93/mdbx-rs/pull/6 is merged (could be soon)

**Comment by michaelsproul** [comment]
This is ready for review and merge (although I understand we may want to prioritise merging Kintsugi)

File: slasher/Cargo.toml (hunk 1 above)
Comment (michaelsproul) [reply slasher/Cargo.toml]: Done

**Review by paulhauner — APPROVED** [review]
Looks great, I couldn't fault it! :rocket:

**Comment by michaelsproul** [comment]
Glorious! Thank you!!

bors r+

**Comment by michaelsproul** [comment]
Fixing up some Rust 1.57.0 lints

**Comment by michaelsproul** [comment]
Let's try again

bors r+

**Pull request closed without merging**